ledger-test-tool/suites/src/main/scala/com/daml/ledger/api/testtool/suites/v1_dev/UpgradingIT.scala

```diff
@@ -0,0 +1,451 @@
+import com.daml.ledger.api.v1.event.CreatedEvent
+import com.daml.ledger.api.v1.transaction_filter.{
+  Filters,
+  InclusiveFilters,
+  TemplateFilter,
+  TransactionFilter,
+}
+import com.daml.ledger.api.v1.transaction_service.GetTransactionsRequest
+import com.daml.ledger.api.v1.value.Identifier.toJavaProto
+import com.daml.ledger.api.v1.value.{Identifier => ScalaPbIdentifier}
+import com.daml.ledger.api.v1.{transaction, value}
+import com.daml.ledger.javaapi.data.codegen.{ContractCompanion, ValueDecoder}
+import com.daml.ledger.javaapi.data.{DamlRecord, Unit => _, _}
+import com.daml.ledger.test.java.upgrade.v1_0_0.upgrade.{UA => UA_V1}
+import com.daml.ledger.test.java.upgrade.v2_0_0.upgrade.{UA => UA_V2, UB => UB_V2}
+import com.daml.ledger.test.java.upgrade.v3_0_0.upgrade.{UB => UB_V3}
+import com.daml.ledger.test.{UpgradeTestDar1_0_0, UpgradeTestDar2_0_0, UpgradeTestDar3_0_0}
+import com.daml.lf.data.Ref
+import com.daml.lf.data.Ref.{PackageName, PackageRef}
+import com.digitalasset.canton.ledger.error.groups.RequestValidationErrors
+
+import java.util.Optional
+import scala.concurrent.{ExecutionContext, Future}
+import scala.jdk.CollectionConverters._
+
+class UpgradingIT extends LedgerTestSuite {
+  implicit val upgradingUA_V1Companion
+      : ContractCompanion.WithoutKey[UA_V1.Contract, UA_V1.ContractId, UA_V1] =
+    UA_V1.COMPANION
+  implicit val upgradingUA_V2Companion
+      : ContractCompanion.WithoutKey[UA_V2.Contract, UA_V2.ContractId, UA_V2] =
+    UA_V2.COMPANION
+  implicit val upgradingUB_V2Companion
+      : ContractCompanion.WithoutKey[UB_V2.Contract, UB_V2.ContractId, UB_V2] =
+    UB_V2.COMPANION
+  implicit val upgradingUB_V3Companion
+      : ContractCompanion.WithoutKey[UB_V3.Contract, UB_V3.ContractId, UB_V3] =
+    UB_V3.COMPANION
+
+  private val PkgRef = PackageRef.Name(PackageName.assertFromString("upgrade-tests"))
+  private val UA_Identifier = ScalaPbIdentifier
+    .fromJavaProto(UA_V1.TEMPLATE_ID.toProto)
+    .withPackageId(PkgRef.toString)
+  private val UB_Identifier = ScalaPbIdentifier
+    .fromJavaProto(UB_V2.TEMPLATE_ID.toProto)
+    .withPackageId(PkgRef.toString)
+
+  private val UnknownPackageNameIdentifier = ScalaPbIdentifier.of(
+    Ref.PackageRef.Name(Ref.PackageName.assertFromString("unknown")).toString,
+    "module",
+    "entity",
+  )
+
+  private val PkgRefUA_V1 =
+    PackageRef.Id(Ref.PackageId.assertFromString(UA_V1.TEMPLATE_ID.getPackageId))
+
+  test(
+    "UDynamicTemplates",
+    "Template-id resolution is updated on package upload during ongoing subscriptions",
+    allocate(SingleParty),
+  )(implicit ec => { case Participants(Participant(ledger, party)) =>
+    for {
+      _ <- `assert subscriptions for unknown package-names fail`(ledger, party)
+
+      // Upload 1.0.0 package
+      _ <- upload(ledger, UpgradeTestDar1_0_0.path)
+
+      _ <- `assert subscriptions for unknown template names succeed if package-name present`(
```

Comment:
It may be better to fail in this situation (the dar should always be loaded before the app).

Reply:
Alright. I'll comment out this test and add a failure condition in a subsequent PR